README for writing a config.yaml (#425)
P2GX/PhenoXtract: phenoxtract/README.md

TOML, or JSON** formats. For an explanation of how to write a config.yaml, see here: [YAML_README](CONFIG_YAML_README.md).

<!-- TOC -->
* [PhenoXtract](#phenoxtract)
  * [How PhenoXtract works](#how-phenoxtract-works)
  * [What format does PhenoXtract expect data to be in?](#what-format-does-phenoxtract-expect-data-to-be-in)
  * [Running PhenoXtract in Rust](#running-phenoxtract-in-rust)
  * [Extracting Individual Data](#extracting-individual-data)
  * [Extracting Phenotypes](#extracting-phenotypes)
  * [Extracting Diseases](#extracting-diseases)
  * [Extracting Interpretations](#extracting-interpretations)
  * [Extracting Measurements](#extracting-measurements)
  * [Extracting Medical Actions](#extracting-medical-actions)
  * [Contexts](#contexts)
  * [Strategies](#strategies)
  * [Authors](#authors)
<!-- TOC -->

## How PhenoXtract works

PhenoXtract begins by extracting the data sources into a [Polars](https://docs.rs/polars/latest/polars/) Dataframes. In
the config file, the user will have specified which Phenopacket elements each column of the data corresponds to. This is

> done by providing a "Series Context" for each column. See [Contexts](#contexts)

Reviewer: ```suggestion
done by providing a `SeriesContext` for each column. See [Contexts](#contexts)
```

and [series_contexts](CONFIG_YAML_README.md#series_contexts) for more information on Series Contexts.

> Once the data has been extracted, "Strategies" are applied, which transform the data into a format that the user can

Reviewer: ```suggestion
Once the data has been extracted, `Strategies` are applied, which transform the data into a format that the application can
```

Finally, the Phenopackets are loaded to .json files in a directory of the user's choice.

## What format does PhenoXtract expect data to be in?

Before the Collection stage of the program, the data must be in a certain format so that it can be understood by
PhenoXtract. How each column should look will be explained in the sections:

* [Extracting Individual Data](#extracting-individual-data)
* [Extracting Phenotypes](#extracting-phenotypes)
* [Extracting Diseases](#extracting-diseases)
* [Extracting Interpretations](#extracting-interpretations)
* [Extracting Measurements](#extracting-measurements)
* [Extracting Medical Actions](#extracting-medical-actions)

## Running PhenoXtract in Rust

Once a config file has been written (see [YAML_README](CONFIG_YAML_README.md) for information on how to write a config.yaml),
PhenoXtract can be run as follows:

```rust
use std::path::PathBuf;
use phenoxtract::phenoxtract::Phenoxtract;

fn test_i_data() {
    let config_path = PathBuf::from(
        "path/to/config.yaml",
    );
    let mut phenoxtract = Phenoxtract::try_from(config_path).unwrap();

>     phenoxtract.run().unwrap();

Reviewer: ```suggestion
    phenoxtract.run()?;
```

(TODO)

## Extracting Medical Actions

(TODO)

## Contexts

In order for PhenoXtract to understand what is inside a column, the user must specify a "Series Context" for that
column. For each Series Context, the user can specify a `header_context`, which describes what is in the header of the
column, and a `data_context` which describes what is in the cells of the column. How one configures a Series Context for
a column (or multiple) is described in [YAML_README](CONFIG_YAML_README.md).

Here is the list of possible values that `header_context` or `data_context` can take:

**Individual data**

- SubjectId
- SubjectSex
- DateOfBirth
- VitalStatus
- TimeAtLastEncounter(TimeElementType)
- TimeOfDeath(TimeElementType)
- CauseOfDeath
- SurvivalTimeDays

**Phenotypes and Disease**

- Hpo
- Disease
- MultiHpoId
- Onset(TimeElementType)

**Genetics**

- Hgvs
- Hgnc

**Measurements**

- QuantitativeMeasurement (assay_id: String, unit_ontology_id: String)
- QualitativeMeasurement (assay_id: String)
- TimeOfMeasurement(TimeElementType)
- ReferenceRange(Boundary)

**Medical Actions**

- TreatmentTarget
- TreatmentIntent
- ResponseToTreatment
- TreatmentTerminationReason

> - ProcedureLabelOrId

Reviewer: Dont forget to change this, when you change the context.
